### PR TITLE
Forgiveness api changes

### PR DESCRIPTION
**Pull request by kevin-wangzefeng**
**What this PR does / why we need it**:
Splited from #34825 , contains api changes that are needed to implement forgiveness:
1. update toleration api types to support forgiveness, added a new field forgivenessSeconds to indicate the duration of time it tolerates a taint.
2. update taint api types, added a new field to indicate the time the taint is added.

**Which issue this PR fixes** : 
Related issue: #1574
Related PR: #34825 

**Special notes for your reviewer**:

**Release note**:

```release-note
forgiveness alpha version api definition
```

### CODE REVIEW

File: pkg/apis/meta/v1/well_known_labels.go
Comment (aveshagarwal) [inline pkg/apis/meta/v1/well_known_labels.go]: Wondering if these changes, TaintNodeNotReady and TaintNodeUnreachable, should be part of this PR, as functionality wise, they are not used in this PR.

**Comment by kevin-wangzefeng** [comment]
@k8s-bot bazel test this

**Comment by kevin-wangzefeng** [comment]
@k8s-bot verify test this

File: pkg/api/types.go
Comment (timothysc) [inline pkg/api/types.go]: Why aren't we using one of our time structs here?  There are now a couple. 
/cc @gmarek

File: pkg/api/validation/validation.go
Comment (timothysc) [inline pkg/api/validation/validation.go]: All the reason to use a time struct.

File: pkg/api/types.go
Comment (gmarek) [inline pkg/api/types.go]: I'm not sure it'll be useful - when making eviction decisions we won't be able to depend on this timestamp either way (clock skew). OTOH it's probably useful for usability perspective, so we can show user when given taint was added, but then comment is wrong (it has the same sense for NoExecute taint as for other ones). @davidopp

File: pkg/api/types.go
Comment (gmarek) [inline pkg/api/types.go]: Yeah - I certainly don't like int representing time. `metav1.Time` is better (for toleration time I don't think microsecond precision makes any sense).

Comment (gmarek) [inline pkg/api/types.go]: Scratch that - it seems I'm still tried. It should be a `time.Duration`.

File: pkg/api/validation/validation.go
Comment (gmarek) [inline pkg/api/validation/validation.go]: `time.Duration` can be negative as well, so this check won't simply go away. The only doubt I have is whether we need to do this check at all. I don't know what our validation policies are - do we reject only things that would break the system in some way (e.g. when we're dividing by given value, or sth.), or we're trying hard to prevent user from doing stupid things. In this case passing negative toleration time wouldn't break anything - it'll be effectively treated as 0 by the system.

**Comment by gmarek** [comment]
@kevin-wangzefeng please - there are a couple comments.

**Comment by gmarek** [comment]
Plus I thought that it was supposed to be moved as a field to beta (whatever this means). @davidopp @timothysc

File: pkg/api/types.go
Comment (gmarek) [inline pkg/api/types.go]: I remember we were discussing it already, but why can't this be a plain value (instead of ptr), with the convention that negative values mean infinity? For backward compatibility we only care that 0 value has a good default semantics (I'm not sure if it's "don't tolerate at all" or "tolerate forever")

Comment (gmarek) [inline pkg/api/types.go]: Now that I looked at our API code it seems so 'XSeconds' is the way we do things. I think it's reasonable to leave it as it is.

**Comment by gmarek** [comment]
Are we bumping generation when changing taints, tolerations? Number of things would be easier if we would. @kevin-wangzefeng @davidopp @erictune

**Comment by davidopp** [comment]
I don't remember seeing any code to bump generation when I reviewed the initial PR.

**Comment by gmarek** [comment]
Originally we were adding/removing annotations, now, as we're going to move Taints/Tolerations to Spec bumping Generation will be needed, as IIUC it's supposed to signalize when we change Spec version.

**Comment by davidopp** [comment]
Yes I think that's a good point. Can you make sure this gets done?

File: pkg/api/validation/validation.go
Comment (kevin-wangzefeng) [inline pkg/api/validation/validation.go]: make sense to me, updated API types as well as validation correspondingly.

**Comment by gmarek** [comment]
Reviewed 1 of 16 files at r1, 14 of 15 files at r2, 4 of 15 files at r3, 10 of 10 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/api/types.go, line 1749 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469#-K_idlV1SCVmkUg6ScWM-r2-1749:-KbLRFQ-AqTYeB5vc4Fo:b-yk54lv) ([raw file](https://github.com/kubernetes/kubernetes/blob/5f3bbdfaa7d7841e52b405be57521fbb9c9f86da/pkg/api/types.go#L1749)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

I'm not sure it'll be useful - when making eviction decisions we won't be able to depend on this timestamp either way (clock skew). OTOH it's probably useful for usability perspective, so we can show user when given taint was added, but then comment is wrong (it has the same sense for NoExecute taint as for other ones). @davidopp @timothysc 
</blockquote></details>

Can you change the comment that this is only for user convenience and might not be precise in case of the clock skew?

---

*[pkg/apis/meta/v1/well_known_labels.go, line 46 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469#-K_idlV2rGOM9r8cfrrG-r1-46:-KbLRAomqlRoFlT3F626:b5uynkm) ([raw file](https://github.com/kubernetes/kubernetes/blob/0feab992789678c212c27eedd7f02b25af433f78/pkg/apis/meta/v1/well_known_labels.go#L46)):*
<details><summary><i>Previously, aveshagarwal (Avesh Agarwal) wrote…</i></summary><blockquote>

Wondering if these changes, TaintNodeNotReady and TaintNodeUnreachable, should be part of this PR, as functionality wise, they are not used in this PR.
</blockquote></details>

What's this?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469)*

**Comment by gmarek** [comment]
Only minor things - mostly LGTM, but I'll let @davidopp to take a final look.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469#-:-KbLRjpjL3Du5NGAfgVg:bk9nd55)*

**Comment by timothysc** [comment]
could you squash to 2 commits please.

**Comment by gmarek** [comment]
@davidopp @timothysc - generation handling should be added when we move Tolerations to filed. (I'm not clear whether Taints are Spec or Status (I guess Spec, but I'm not sure), so I don't know about those).

File: pkg/api/types.go
Comment (davidopp) [inline pkg/api/types.go]: @gmarek I think you said this was going to be a new controller, not NodeController?

File: pkg/api/types.go
Comment (davidopp) [inline pkg/api/types.go]: s/tolerate the taint no time/evict immediately/

File: pkg/api/types.go
Comment (davidopp) [inline pkg/api/types.go]: Is TimeAdded only written for NoExecute taints? If so, I would mention that.
Also, if the actual implementation ignores this field (as @gmarek is suggesting), then maybe you should remove most of the text and just say "TimeAdded represents the time at which the taint was added." And if it's only written for NoExecute taint, then also say "It is only written for NoExecute taints."

File: pkg/api/types.go
Comment (gmarek) [inline pkg/api/types.go]: After some though I figured out that reusing NodeControllers eviction logic makes no sense for Taints (it has rate limiting which we don't want for taint based evictions). So I just added the 'child controller' in the same way CIDR allocator is.

I don't have a strong opinion on this - it was just easier to write, as I didn't need to write all the boilerplate. Plus NC will strongly depend on TaintController to do evictions, so we may want to keep them together for now.

**Comment by kevin-wangzefeng** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/api/types.go, line 1749 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469#-K_idlV1SCVmkUg6ScWM-r2-1749:-KbQ7CeokPoO3rFhOmml:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5f3bbdfaa7d7841e52b405be57521fbb9c9f86da/pkg/api/types.go#L1749)):*
<details><summary><i>Previously, davidopp (David Oppenheimer) wrote…</i></summary><blockquote>

Is TimeAdded only written for NoExecute taints? If so, I would mention that.
Also, if the actual implementation ignores this field (as @gmarek is suggesting), then maybe you should remove most of the text and just say "TimeAdded represents the time at which the taint was added." And if it's only written for NoExecute taint, then also say "It is only written for NoExecute taints."

</blockquote></details>

Done.

---

*[pkg/api/types.go, line 1845 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469#-K_idlV1SCVmkUg6ScWM-r4-1845:-KbQ7_y5DXi18YPlty3L:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/f1943f4f372874fe32ebd0c0476323639849298b/pkg/api/types.go#L1845)):*
<details><summary><i>Previously, davidopp (David Oppenheimer) wrote…</i></summary><blockquote>

s/tolerate the taint no time/evict immediately/
</blockquote></details>

Done.

---

*[pkg/apis/meta/v1/well_known_labels.go, line 46 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469#-K_idlV2rGOM9r8cfrrG-r1-46:-KbQ7thN403Sd4W24D3U:bieeac5) ([raw file](https://github.com/kubernetes/kubernetes/blob/0feab992789678c212c27eedd7f02b25af433f78/pkg/apis/meta/v1/well_known_labels.go#L46)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

What's this?
</blockquote></details>

It 's the taint keys for nodeNotReady and NodeUnreachable. Since they are more related to the flag `--use-taint-based-evictions`, may better be a part of the taint controller PR(#40355).

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469)*

**Comment by kevin-wangzefeng** [comment]
Refreshed and squashed, PTAL.

**Comment by gmarek** [comment]
One nit. But we also need an approval with the powers here..

---

Reviewed 1 of 15 files at r3, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/api/types.go, line 1749 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469#-K_idlV1SCVmkUg6ScWM-r2-1749:-KbQqgxy8HFR7XmIBpwr:b-lyu8ql) ([raw file](https://github.com/kubernetes/kubernetes/blob/5f3bbdfaa7d7841e52b405be57521fbb9c9f86da/pkg/api/types.go#L1749)):*
<details><summary><i>Previously, kevin-wangzefeng (Kevin) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

How much work it'd be to write this for other taints as well? If a lot, then it's OK, but I think it'd be nice to have consistency.

---

*[pkg/apis/meta/v1/well_known_labels.go, line 46 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469#-K_idlV2rGOM9r8cfrrG-r1-46:-KbQr4oqwqh59EW5ZZCr:b-bn3q8o) ([raw file](https://github.com/kubernetes/kubernetes/blob/0feab992789678c212c27eedd7f02b25af433f78/pkg/apis/meta/v1/well_known_labels.go#L46)):*
<details><summary><i>Previously, kevin-wangzefeng (Kevin) wrote…</i></summary><blockquote>

It 's the taint keys for nodeNotReady and NodeUnreachable. Since they are more related to the flag `--use-taint-based-evictions`, may better be a part of the taint controller PR(#40355).
</blockquote></details>

Meh, I'm fine with that being here.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469#-:-KbQrGWDRv6sn2cW8HoC:b-mo6vd1)*

**Comment by gmarek** [comment]
someone with approval powers*

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469#-:-KbQrRq90V1SG6T3T6Gz:b-hq3n96)*

**Comment by davidopp** [comment]
Let me know when it's ready to approve and I will do it. (Technically I'm not an approver, because the approvers files haven't really been set up properly yet, but for now people who used to be able to add LGTM labels can also approve anything.)

**Review by timothysc — APPROVED** [review]
/lgtm

**Comment by timothysc** [comment]
@davidopp you just need to apply the "/approve" label.  The whole 'lgtm+approve' thing seems silly to me ( /cc @kubernetes/sig-contributor-experience-bugs )

**Comment by timothysc** [comment]
@k8s-bot kops aws e2e test this

**Comment by davidopp** [comment]
LGTM added

**Comment by kevin-wangzefeng** [comment]
Review status: 14 of 15 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/apis/meta/v1/well_known_labels.go, line 46 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469#-K_idlV2rGOM9r8cfrrG-r1-46:-KbV0lQVcqDYSbBWlltE:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/0feab992789678c212c27eedd7f02b25af433f78/pkg/apis/meta/v1/well_known_labels.go#L46)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

Meh, I'm fine with that being here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39469)*

**Comment by kevin-wangzefeng** [comment]
added definitions of taint keys for nodeNotReady and NodeUnreachable into this PR, PTAL.

**Comment by kevin-wangzefeng** [comment]
@k8s-bot unit test this

**Comment by kevin-wangzefeng** [comment]
@k8s-bot gci gce e2e test this

**Comment by kevin-wangzefeng** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by gmarek** [comment]
Great - it's in!

File: pkg/api/v1/generated.proto
Comment (davidopp) [inline pkg/api/v1/generated.proto]: Stupid question - why is this int64 instead of *int64? Is it some weirdness of the auto-generation?

Comment (gmarek) [inline pkg/api/v1/generated.proto]: I guess 'optional' means it's a pointer, doesn't it?

**Pull request merged**
